ceph: add an option to enable/disable merge all placement

subhamkrai opened the pull request:
earlier, we had the issue for osd placement regarding
whether we should merge the placements or not.

**Description of your changes:**
Now, we have option `skipApplyAllPlacement` to
enable/disable osd placement. By default it is false
which means it will merge both placement.All() and
storageClassDeviceSets.Placement, when true
it will only apply storageClassDeviceSets.Placement.

Also, adding unit test.

Closes: https://github.com/rook/rook/issues/8135
Signed-off-by: subhamkrai <srai@redhat.com>

<!-- Please take a look at our [Contributing](https://rook.io/docs/rook/master/development-flow.html)
documentation before submitting a Pull Request!
Thank you for contributing to Rook! -->

**Which issue is resolved by this Pull Request:**
Resolves #8135 

**Checklist:**

- [ ] **Commit Message Formatting**: Commit titles and messages follow guidelines in the [developer guide](https://rook.io/docs/rook/master/development-flow.html#commit-structure).
- [ ] **Skip Tests for Docs**: Add the flag for skipping the build if this is only a documentation change. See [here](https://github.com/rook/rook/blob/master/INSTALL.md#skip-ci) for the flag.
- [ ] **Skip Unrelated Tests**: Add a flag to run tests for a specific storage provider. See [test options](https://github.com/rook/rook/blob/master/INSTALL.md#test-storage-provider).
- [ ] Reviewed the developer guide on [Submitting a Pull Request](https://rook.io/docs/rook/master/development-flow.html#submitting-a-pull-request)
- [ ] Documentation has been updated, if necessary.
- [ ] Unit tests have been added, if necessary.
- [ ] Integration tests have been added, if necessary.
- [ ] Pending release notes updated with breaking and/or notable changes, if necessary.
- [ ] Upgrade from previous release is tested and upgrade user guide is updated, if necessary.
- [ ] Code generation (`make codegen`) has been run to update object specifications, if necessary.

subhamkrai commented:
@travisn do we want this option for non-PVCs clusters too as for non-PVCs clusters we use this method https://github.com/rook/rook/blob/95505592c8142a82095892fae0dfbd6de9bd0497/pkg/apis/ceph.rook.io/v1/placement.go#L134-L136 which has `merge()` call which is overriding the placement with applied placement.

We can edit the `GetOSDPlacement` but his method is used by other files too or create a new function for non-PVCs similar to the PVC, thoughts?

sp98 commented on cluster/charts/rook-ceph/templates/resources.yaml:
```diff
@@ -2122,6 +2122,8 @@ spec:
                         type: object
                       nullable: true
                       type: array
+                    skipApplyAllPlacement:
```
New field should be updated in `ceph-cluster-crd.md` file.

subhamkrai replied:
added

sp98 commented on pkg/operator/ceph/cluster/osd/spec.go:
```diff
@@ -715,14 +715,10 @@ func (c *Cluster) makeDeployment(osdProps osdProperties, osd OSDInfo, provisionC
-		// the "all" placement is applied separately so it will have lower priority.
-		// We want placement from the storageClassDeviceSet to be applied and override
-		// the "all" placement if there are any overlapping placement settings.
-		c.spec.Placement.All().ApplyToPodSpec(&deployment.Spec.Template.Spec)
```
we should keep the comments and modify them if required. They look useful.

sp98 commented on pkg/operator/ceph/cluster/osd/provision_spec.go:
```diff
@@ -157,13 +157,9 @@ func (c *Cluster) provisionPodTemplateSpec(osdProps osdProperties, restart v1.Re
 		podSpec.DNSPolicy = v1.DNSClusterFirstWithHostNet
 	}
 	if osdProps.onPVC() {
-		// The "all" placement is applied separately so it will have lower priority.
```
we should keep the comments and modify them if required. They look useful.

sp98 commented on pkg/operator/ceph/cluster/osd/provision_spec.go:
```diff
@@ -157,13 +157,9 @@ func (c *Cluster) provisionPodTemplateSpec(osdProps osdProperties, restart v1.Re
-		// Apply storageClassDeviceSet PreparePlacement
-		// If nodeAffinity is specified both in the device set and "all" placement,
-		// they will be merged.
+		if !c.spec.Storage.SkipApplyAllPlacement {
```
Took me a while to comprehend this condition. Maybe its the NOT operator on an already negative filter, or maybe its just me!

sp98 reviewed (changes requested):
Are there any unit tests for validating the placements? I didn't see any.

leseb commented on cluster/examples/kubernetes/ceph/cluster-on-pvc.yaml:
```diff
@@ -47,6 +47,8 @@ spec:
     disable: false
   storage:
+    # when skipApplyAllPlacement false, will merge both placement.All() and storageClassDeviceSets.Placement.
+    skipApplyAllPlacement: true
```
Is this a setting we intend the user to change if we set it to true for the PVC scenario exclusively? If so, when? If can't we internally determine this and merge when necessary?

subhamkrai commented on pkg/operator/ceph/cluster/osd/provision_spec.go:
```diff
@@ -157,17 +157,31 @@ func (c *Cluster) provisionPodTemplateSpec(osdProps osdProperties, restart v1.Re
-		// they will be merged.
+		// when "onlyApplySpecificOSDPlacement" false, in case of NodeAffinity it will merge both spec.placement.All() and storageClassDeviceSets.preparePlacement,
+		// in case of other preparePlacement rule like PodAffinity, PodAntiAffinity... it will override last prepareplacement with the current preparePlacement applied,
+		// technically in another word for prepareplacement apart from NodeAffinity, it will same as "onlyApplySpecificOSDPlacement" true as storageClassDeviceSets.preparePlacement
+		// will override spec.placement.All(). See ApplyToPodSpec().
+
+		// when "onlyApplySpecificOSDPlacement" true, it will only apply storageClassDeviceSets.preparePlacement
+		if !c.spec.Storage.OnlyApplySpecificOSDPlacement {
+			// apply spec.placement.All()
+			c.spec.Placement.All().ApplyToPodSpec(&podSpec)
+		}
+		// apply storageClassDeviceSets.preparePlacement
 		osdProps.getPreparePlacement().ApplyToPodSpec(&podSpec)
 	} else {
-		p := cephv1.GetOSDPlacement(c.spec.Placement)
-		p.ApplyToPodSpec(&podSpec)
+		// when "onlyApplySpecificOSDPlacement" false, in case of NodeAffinity it will merge both spec.placement.All() and spec.Placement.prepareosd,
+		// in case of other preparePlacement rule like PodAffinity, PodAntiAffinity... it will override last preparePlacement with the current preparePlacement applied,
+		// technically in another word for preparePlacement apart from NodeAffinity, it will same as "onlyApplySpecificOSDPlacement" true as spec.Placement.prepareosd will
+		// override spec.placement.All(). See ApplyToPodSpec().
+
+		// when "onlyApplySpecificOSDPlacement" true, it will only apply spec.placement.prepareosd
+		if !c.spec.Storage.OnlyApplySpecificOSDPlacement {
+			// apply spec.placement.All()
+			c.spec.Placement.All().ApplyToPodSpec(&podSpec)
+		}
+		// apply spec.placement.prepareosd
+		c.spec.Placement[cephv1.KeyOSDPrepare].ApplyToPodSpec(&podSpec)
```
@sp98 @travisn can you give it a try to see if it is clear... I tried to make it as simple as possible.

And, also I have not tested my changes, will do once the approach is verified.
The reason for not testing is I'm not using `GetOSDPlacement()` as it will make it more complex and doing similar as pvc case.

subhamkrai commented on pkg/apis/ceph.rook.io/v1/types.go:
```diff
@@ -1890,6 +1890,8 @@ type StorageScopeSpec struct {
 	// +optional
 	UseAllNodes bool `json:"useAllNodes,omitempty"`
+	// +optional
+	SkipApplyAllPlacement bool `json:"skipApplyAllPlacement,omitempty"`
```
done.

subhamkrai commented:
> The testing will also be critical to ensure we don't have a regression. Can you comment on what you observe in manual tests?

I did some testing for PVC case,

when `onlyApplySpecificOSDPlacement:  true`

`kubectl get cephclusters.ceph.rook.io -o yaml`

```
placement:
      all:
        nodeAffinity:
          requiredDuringSchedulingIgnoredDuringExecution:
            nodeSelectorTerms:
            - matchExpressions:
              - key: failure-domain.beta.kubernetes.io/zone
                operator: In
                values:
                - ap-southeast-1a
                - ap-southeast-1b
                - ap-southeast-1c
    security:
      kms: {}
    storage:
      onlyApplySpecificOSDPlacement: true
      storageClassDeviceSets:
      - count: 3
        name: set1
        placement:
          nodeAffinity:
            requiredDuringSchedulingIgnoredDuringExecution:
              nodeSelectorTerms:
              - matchExpressions:
                - key: topology.kubernetes.io/zone
                  operator: In
                  values:
                  - ap-southeast-1b
                  - ap-southeast-1a
```
we can see only `storageClassDeviceSets` placement is applied.

```
kubectl get pod rook-ceph-osd-1-5984794c49-btw9w -o yaml
```

```
  uid: 2979dc79-8eaf-4f33-b190-2d55ba87ddc2
spec:
  affinity:
    nodeAffinity:
      requiredDuringSchedulingIgnoredDuringExecution:
        nodeSelectorTerms:
        - matchExpressions:
          - key: topology.kubernetes.io/zone
            operator: In
            values:
            - ap-southeast-1b
            - ap-southeast-1a
          - key: topology.kubernetes.io/zone
            operator: In
            values:
            - ap-southeast-1b
  containers:
```
when `onlyApplySpecificOSDPlacement:  true`

```
kubectl get cephclusters.ceph.rook.io -o yaml
```
```
placement:
      all:
        nodeAffinity:
          requiredDuringSchedulingIgnoredDuringExecution:
            nodeSelectorTerms:
            - matchExpressions:
              - key: failure-domain.beta.kubernetes.io/zone
                operator: In
                values:
                - ap-southeast-1a
                - ap-southeast-1b
                - ap-southeast-1c
    security:
      kms: {}
    storage:
      storageClassDeviceSets:
      - count: 3
        name: set1
        placement:
          nodeAffinity:
            requiredDuringSchedulingIgnoredDuringExecution:
              nodeSelectorTerms:
              - matchExpressions:
                - key: topology.kubernetes.io/zone
                  operator: In
                  values:
                  - ap-southeast-1b
                  - ap-southeast-1a
```
We can see both placements applied
```
kubectl get pod rook-ceph-osd-0-7b7b5cff6-s2jbx -o yaml
```
```
uid: 26b73477-b8c5-4ccc-a915-dd9fb86f6ccc
spec:
  affinity:
    nodeAffinity:
      requiredDuringSchedulingIgnoredDuringExecution:
        nodeSelectorTerms:
        - matchExpressions:
          - key: failure-domain.beta.kubernetes.io/zone
            operator: In
            values:
            - ap-southeast-1a
            - ap-southeast-1b
            - ap-southeast-1c
          - key: topology.kubernetes.io/zone
            operator: In
            values:
            - ap-southeast-1b
            - ap-southeast-1a
          - key: topology.kubernetes.io/zone
            operator: In
            values:
            - ap-southeast-1b
  containers:
```

is this testing good for PVC case? @travisn @sp98

sp98 commented on cluster/examples/kubernetes/ceph/cluster-on-local-pvc.yaml:
```diff
@@ -231,6 +231,8 @@ spec:
               accessModes:
                 - ReadWriteOnce
+    # when onlyApplySpecificOSDPlacement false, will merge both placement.All() and storageClassDeviceSets.Placement
+    onlyApplySpecificOSDPlacement: false
```
I believe the default behavior that we are trying to imply is to keep `onlyApplySpecificOSDPlacement: false`

So should we comment this field out by default in all the example CRs?  Users can remove the comment and mark it as true if don't want to merge. 

@travisn  @subhamkrai

sp98 commented on cluster/examples/kubernetes/ceph/cluster-on-local-pvc.yaml:
```diff
@@ -231,6 +231,8 @@ spec:
               volumeMode: Block
               accessModes:
                 - ReadWriteOnce
+    # when onlyApplySpecificOSDPlacement false, will merge both placement.All() and storageClassDeviceSets.Placement
```
I would suggest something like 
```suggestion
    # onlyApplySpecificOSDPlacement  will only use `placement` or  `preparePlacement` under `storageClassDeviceSets` when set to true.  In case of false,  `placement.All` will be merged with `placement` or  `preparePlacement` under `storageClassDeviceSets`
```

sp98 commented on cluster/examples/kubernetes/ceph/cluster-on-pvc.yaml:
```diff
@@ -47,6 +47,8 @@ spec:
   crashCollector:
     disable: false
   storage:
+    # when onlyApplySpecificOSDPlacement false, will merge both placement.All() and storageClassDeviceSets.Placement.
```
I would suggest something like:

```suggestion
    #  onlyApplySpecificOSDPlacement  will only use `placement` or  `preparePlacement` under `storageClassDeviceSets` when set to true.  In case of false,  `placement.All` will be merged with `placement` or  `preparePlacement` under `storageClassDeviceSets`
```

sp98 commented on cluster/examples/kubernetes/ceph/cluster-on-local-pvc.yaml:
```diff
@@ -231,6 +231,8 @@ spec:
               accessModes:
                 - ReadWriteOnce
+    # when onlyApplySpecificOSDPlacement false, will merge both placement.All() and storageClassDeviceSets.Placement
+    onlyApplySpecificOSDPlacement: false
```
should we add  this field in osd on disks example CRs as well?

subhamkrai replied:
added.

subhamkrai commented:
> Yes that looks good. But for the second example did you mean `onlyApplySpecificOSDPlacement: false`?

right :smile:

subhamkrai commented on pkg/operator/ceph/cluster/osd/provision_spec.go:
```diff
@@ -157,17 +157,31 @@ func (c *Cluster) provisionPodTemplateSpec(osdProps osdProperties, restart v1.Re
+			c.spec.Placement.All().ApplyToPodSpec(&podSpec)
+		}
+		// apply storageClassDeviceSets.preparePlacement
 		osdProps.getPreparePlacement().ApplyToPodSpec(&podSpec)
```
done.

subhamkrai commented:
> Are there any unit tests for validating the placements? I didn't see any.

added unit for placements. 

Also updated comments by merging @travisn comments, PTAL

sp98 commented on Documentation/ceph-cluster-crd.md:
```diff
@@ -216,6 +218,7 @@ For more details on the mons and when to choose a number other than `3`, see the
   * `config`: Config settings applied to all OSDs on the node unless overridden by `devices`. See the [config settings](#osd-configuration-settings) below.
   * [storage selection settings](#storage-selection-settings)
   * [Storage Class Device Sets](#storage-class-device-sets)
+  * `onlyApplySpecificOSDPlacement`: `false` or `true`, In case of false, for OSD, will merge nodeAffinity of the placement.All() with placement.osd(in case of non-PVCs) or storageClassDeviceSets.placement(in case of PVCs), for prepareOSD will merge nodeAffinity of the placement.All() with placement.prepareosd(in case of non-PVCs) or storageClassDeviceSets.preparePlacement(in case of PVCs) vice-v ...
```
I would suggest using an easier explanation as mentioned in the [comment](https://github.com/rook/rook/pull/8381#discussion_r678411565).

subhamkrai replied:
added.

sp98 commented on cluster/examples/kubernetes/ceph/cluster.yaml:
```diff
@@ -229,6 +229,8 @@ spec:
 #      config: # configuration can be specified at the node level which overrides the cluster level config
 #    - name: "172.17.4.301"
 #      deviceFilter: "^sd."
+    # when onlyApplySpecificOSDPlacement false, will merge both placement.All() and storageClassDeviceSets.Placement.
```
same comment `when onlyApplySpecificOSDPlacement false, will merge both placement.All() and storageClassDeviceSets.Placement.` is used for both cluster.yaml and cluster-on-pvc.yaml.   I think we need to differentiate between osd-on-pvc and osd-on-disks scenarios here.

subhamkrai replied:
updated.

sp98 commented on pkg/operator/ceph/cluster/osd/spec_test.go:
```diff
@@ -707,3 +708,164 @@ func getDummyDeploymentOnNode(clientset *fake.Clientset, c *Cluster, nodeName st
+		DataPathMap: opconfig.NewDatalessDaemonDataPathMap(c.clusterInfo.Namespace, "/var/lib/rook"),
+	}
+
+	// for OSD daemon
```
lets start every comment with capital letter to be consistent with other areas in the code base.

subhamkrai commented on Documentation/ceph-cluster-crd.md:
```diff
@@ -216,7 +218,12 @@ For more details on the mons and when to choose a number other than `3`, see the
   * [storage selection settings](#storage-selection-settings)
   * [Storage Class Device Sets](#storage-class-device-sets)
-* `disruptionManagement`: The section for configuring management of daemon disruptions
+  * `onlyApplySpecificOSDPlacement`: `false` or `true`, The placement for OSDs is computed from several different places:
```
done.

subhamkrai commented on pkg/apis/ceph.rook.io/v1/types.go:
```diff
@@ -1890,6 +1890,8 @@ type StorageScopeSpec struct {
 	// +optional
 	UseAllNodes bool `json:"useAllNodes,omitempty"`
+	// +optional
+	OnlyApplySpecificOSDPlacement bool `json:"onlyApplySpecificOSDPlacement,omitempty"`
```
updated.

subhamkrai commented on pkg/operator/ceph/cluster/osd/spec.go:
```diff
@@ -715,17 +715,13 @@ func (c *Cluster) makeDeployment(osdProps osdProperties, osd OSDInfo, provisionC
-		// apply storageClassDeviceSet Placement
-		// If nodeAffinity is specified both in the device set and "all" placement,
-		// they will be merged.
+		c.placementForOsdandPrepareOsdPods(&deployment.Spec.Template.Spec)
```
I don't think so, for PVCs-based osd it is of cephv1.Placement types and for non-PVCs, it is of cephv1.PlacementSpec

subhamkrai commented on pkg/operator/ceph/cluster/osd/spec.go:
```diff
@@ -715,17 +715,13 @@ func (c *Cluster) makeDeployment(osdProps osdProperties, osd OSDInfo, provisionC
 	} else {
-		p := cephv1.GetOSDPlacement(c.spec.Placement)
-		p.ApplyToPodSpec(&deployment.Spec.Template.Spec)
+		c.placementForOsdandPrepareOsdPods(&deployment.Spec.Template.Spec)
```
not changing as types for PVCs OSD and non-PVCs OSD are different

subhamkrai commented on pkg/operator/ceph/cluster/osd/spec.go:
```diff
@@ -745,6 +741,22 @@ func (c *Cluster) makeDeployment(osdProps osdProperties, osd OSDInfo, provisionC
 }
 
+// applyOSDPlacement apply spec.placement.all if OnlyApplyOSDPlacement set to false
+func (c *Cluster) applyOSDPlacement(d *v1.PodSpec) {
```
done

hjkatz commented on Documentation/ceph-cluster-crd.md:
```diff
@@ -216,7 +218,9 @@ For more details on the mons and when to choose a number other than `3`, see the
   * `config`: Config settings applied to all OSDs on the node unless overridden by `devices`. See the [config settings](#osd-configuration-settings) below.
   * [storage selection settings](#storage-selection-settings)
   * [Storage Class Device Sets](#storage-class-device-sets)
-* `disruptionManagement`: The section for configuring management of daemon disruptions
```
@travisn Was this an intended deletion from the docs?

hjkatz replied:
I was quite confused while reading the docs and trying to decipher where the `disruptionManagement` key should be placed 😅

subhamkrai replied:
added back. Thanks, @hjkatz
